# Computational Intelligence Homework 2

# Used libraries

In [1]:
!pip install scikit-learn

You should consider upgrading via the 'd:\apps\python3.7.5\python.exe -m pip install --upgrade pip' command.


In [2]:
import pandas as pd
import numpy as np
from tabulate import tabulate
from sklearn.model_selection import KFold
import math

# Fetching the data

In [3]:
trainset = pd.read_csv('../Dataset/train.csv')
testset = pd.read_csv('../Dataset/test.csv')

## Demonstrating the data
the train set

In [4]:
trainset

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

the test set

In [5]:
testset

PassengerId  Pclass                                          Name  \
0            892       3                              Kelly, Mr. James   
1            893       3              Wilkes, Mrs. James (Ellen Needs)   
2            894       2                     Myles, Mr. Thomas Francis   
3            895       3                              Wirz, Mr. Albert   
4            896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)   
..           ...     ...                                           ...   
413         1305       3                            Spector, Mr. Woolf   
414         1306       1                  Oliva y Ocana, Dona. Fermina   
415         1307       3                  Saether, Mr. Simon Sivertsen   
416         1308       3                           Ware, Mr. Frederick   
417         1309       3                      Peter, Master. Michael J   

        Sex   Age  SibSp  Parch              Ticket      Fare Cabin Embarked  
0      male  34.5      0      0              330911    7.8292   NaN        Q  
1    female  47.0      1      0              363272    7.0000   NaN        S  
2      male  62.0      0      0              240276    9.6875   NaN        Q  
3      male  27.0      0      0              315154    8.6625   NaN        S  
4    female  22.0      1      1             3101298   12.2875   NaN        S  
..      ...   ...    ...    ...                 ...       ...   ...      ...  
413    male   NaN      0      0           A.5. 3236    8.0500   NaN        S  
414  female  39.0      0      0            PC 17758  108.9000  C105        C  
415    male  38.5      0      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
416    male   NaN      0      0              359309    8.0500   NaN        S  
417    male   NaN      1      1                2668   22.3583   NaN        C  

[418 rows x 11 columns]

## Showing how much data is missing

In [6]:
trainset.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [7]:
trainset = trainset.fillna(method='pad')
# data imputation, the numerical missing data is replaced by the mean of the dataset and the categorical data is 
# replaced by the mod of the dataset.
trainset.isna().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          1
Embarked       0
dtype: int64

In [8]:
testset.isna().sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
dtype: int64

In [9]:
testset = testset.fillna(method='pad')
# data imputation, the numerical missing data is replaced by the mean of the dataset and the categorical data is 
# replaced by the mod of the dataset.
testset.isna().sum()

PassengerId     0
Pclass          0
Name            0
Sex             0
Age             0
SibSp           0
Parch           0
Ticket          0
Fare            0
Cabin          12
Embarked        0
dtype: int64

# Feature Engineeing, preprocessing our data
removing unnecessary features and keeping the important ones. 

***NOTE: the cabin number had 147 categories and 687 missing data so I decided to remove this feature.***

In [10]:
# removing Name, Ticket and Cabin features.
trainset.drop(['Name', 'Ticket', 'Cabin'], axis=1, inplace=True)
trainset

PassengerId  Survived  Pclass     Sex   Age  SibSp  Parch     Fare  \
0              1         0       3    male  22.0      1      0   7.2500   
1              2         1       1  female  38.0      1      0  71.2833   
2              3         1       3  female  26.0      0      0   7.9250   
3              4         1       1  female  35.0      1      0  53.1000   
4              5         0       3    male  35.0      0      0   8.0500   
..           ...       ...     ...     ...   ...    ...    ...      ...   
886          887         0       2    male  27.0      0      0  13.0000   
887          888         1       1  female  19.0      0      0  30.0000   
888          889         0       3  female  19.0      1      2  23.4500   
889          890         1       1    male  26.0      0      0  30.0000   
890          891         0       3    male  32.0      0      0   7.7500   

    Embarked  
0          S  
1          C  
2          S  
3          S  
4          S  
..       ...  
886        S  
887        S  
888        S  
889        C  
890        Q  

[891 rows x 9 columns]

In [11]:
# removing Name, Ticket and Cabin features.
testset.drop(['Name', 'Ticket', 'Cabin'], axis=1, inplace=True)
testset

PassengerId  Pclass     Sex   Age  SibSp  Parch      Fare Embarked
0            892       3    male  34.5      0      0    7.8292        Q
1            893       3  female  47.0      1      0    7.0000        S
2            894       2    male  62.0      0      0    9.6875        Q
3            895       3    male  27.0      0      0    8.6625        S
4            896       3  female  22.0      1      1   12.2875        S
..           ...     ...     ...   ...    ...    ...       ...      ...
413         1305       3    male  28.0      0      0    8.0500        S
414         1306       1  female  39.0      0      0  108.9000        C
415         1307       3    male  38.5      0      0    7.2500        S
416         1308       3    male  38.5      0      0    8.0500        S
417         1309       3    male  38.5      1      1   22.3583        C

[418 rows x 8 columns]

### changing categorical data to numerical using one hot encoding method.

in below firstly, we counted the number of categories for each column.

In [12]:
print('Sex categories =',trainset['Sex'].unique())
print('Embarked categories =',trainset['Embarked'].unique())

Sex categories = ['male' 'female']
Embarked categories = ['S' 'C' 'Q']


In [13]:
# replacing category columns with boolean columns
sex_column = trainset['Sex'].to_numpy()
male_column = (sex_column == 'male') * 1 # this multiplication converts boolean to integer values
female_column = (sex_column == 'female') * 1 # this multiplication converts boolean to integer values
trainset.insert(3, 'Male', male_column)
trainset.insert(3, 'Female', female_column)
trainset[['Female', 'Male', 'Sex']]

Female  Male     Sex
0         0     1    male
1         1     0  female
2         1     0  female
3         1     0  female
4         0     1    male
..      ...   ...     ...
886       0     1    male
887       1     0  female
888       1     0  female
889       0     1    male
890       0     1    male

[891 rows x 3 columns]

In [14]:
trainset.drop('Sex', axis=1, inplace=True)
trainset

PassengerId  Survived  Pclass  Female  Male   Age  SibSp  Parch     Fare  \
0              1         0       3       0     1  22.0      1      0   7.2500   
1              2         1       1       1     0  38.0      1      0  71.2833   
2              3         1       3       1     0  26.0      0      0   7.9250   
3              4         1       1       1     0  35.0      1      0  53.1000   
4              5         0       3       0     1  35.0      0      0   8.0500   
..           ...       ...     ...     ...   ...   ...    ...    ...      ...   
886          887         0       2       0     1  27.0      0      0  13.0000   
887          888         1       1       1     0  19.0      0      0  30.0000   
888          889         0       3       1     0  19.0      1      2  23.4500   
889          890         1       1       0     1  26.0      0      0  30.0000   
890          891         0       3       0     1  32.0      0      0   7.7500   

    Embarked  
0          S  
1          C  
2          S  
3          S  
4          S  
..       ...  
886        S  
887        S  
888        S  
889        C  
890        Q  

[891 rows x 10 columns]

### doing the same thing for Embarked column

In [15]:
embarked_column = trainset['Embarked'].to_numpy()
s_column = (embarked_column == 'S') * 1 # this multiplication converts boolean to integer values
c_column = (embarked_column == 'C') * 1 # this multiplication converts boolean to integer values
q_column = (embarked_column == 'Q') * 1 # this multiplication converts boolean to integer values
trainset.insert(9, 'Embarked_S', s_column)
trainset.insert(9, 'Embarked_C', c_column)
trainset.insert(9, 'Embarked_Q', q_column)
trainset.drop('Embarked', axis=1, inplace=True)
trainset.drop('PassengerId', axis=1, inplace=True)

### doing the exact same thing for the trainset

In [16]:
# replacing category columns with boolean columns
sex_column = testset['Sex'].to_numpy()
male_column = (sex_column == 'male') * 1 # this multiplication converts boolean to integer values
female_column = (sex_column == 'female') * 1 # this multiplication converts boolean to integer values
testset.insert(3, 'Male', male_column)
testset.insert(3, 'Female', female_column)
testset.drop('Sex', axis=1, inplace=True)

embarked_column = testset['Embarked'].to_numpy()
s_column = (embarked_column == 'S') * 1 # this multiplication converts boolean to integer values
c_column = (embarked_column == 'C') * 1 # this multiplication converts boolean to integer values
q_column = (embarked_column == 'Q') * 1 # this multiplication converts boolean to integer values
testset.insert(8, 'Embarked_S', s_column)
testset.insert(8, 'Embarked_C', c_column)
testset.insert(8, 'Embarked_Q', q_column)
testset.drop('Embarked', axis=1, inplace=True)
testset.drop('PassengerId', axis=1, inplace=True)

### Normalizing Fare and Age Columns

In [17]:
age_max = np.max(trainset['Age'].to_numpy())
age_min = np.min(trainset['Age'].to_numpy())
norm_age = (trainset['Age'].to_numpy() - age_min) / (age_max - age_min)
trainset.insert(4, 'newAge', norm_age)
trainset.drop('Age', axis=1, inplace=True)

fare_max = np.max(trainset['Fare'].to_numpy())
fare_min = np.min(trainset['Fare'].to_numpy())
norm_fare = (trainset['Fare'].to_numpy() - fare_min) / (fare_max - fare_min)
trainset.insert(7, 'newFare', norm_fare)
trainset.drop('Fare', axis=1, inplace=True)

norm_age_test = (testset['Age'].to_numpy() - age_min) / (age_max - age_min)
testset.insert(3, 'newAge', norm_age_test)
testset.drop('Age', axis=1, inplace=True)

norm_fare_test = (testset['Fare'].to_numpy() - fare_min) / (fare_max - fare_min)
testset.insert(6, 'newFare', norm_fare_test)
testset.drop('Fare', axis=1, inplace=True)

# Extracting the output column from trainset

In [18]:
survived_vec = trainset['Survived'].to_numpy()
trainset.drop('Survived', axis=1, inplace=True)
print('final trainset')
print('final testset')

# K-nearest neighbours (KNN) model

In [19]:
# returns the number of true positives
def n_TP(actual, prediction):
    n_true = actual==prediction # boolean list of truly predictions
    n_positive = prediction==1 # boolean list of positive predictions
    return np.sum(np.logical_and(n_true, n_positive) * 1)   # returns the number of true&positives
                                                            # multiplication converts boolean to integer values

# returns the number of true negatives
def n_TN(actual, prediction):
    n_true = actual==prediction # boolean list of truly predictions
    n_negative = prediction==0 # boolean list of negative predictions
    return np.sum(np.logical_and(n_true, n_negative) * 1)   # returns the number of true&negatives
                                                            # multiplication converts boolean to integer values

In [20]:
# chose to use precision because we don't favor false negatives over false positives or vise-versa.
def precision(actual, prediction):
    return (n_TP(actual, prediction) + n_TN(actual, prediction)) / (actual.shape[0])

the distance between the data is defined by their euclidean similarity.

In [21]:
# parameters semantics:
# data is the sample which is considered as the train data, the predictions are based on this data.
# x is the sample given to the model for new prediction.
# k is the hyper parameter of the model. which decides how many neighbours to consider for prediction.
def neighbours_of(data, x, k, y_data):
    
    # initialize nearest neighbours
    neibs = data[0:k]
    y_neibs = y_data[0:k]
    for data_i in range(k, data.shape[0]):
        
        # reset max index each iteration
        max_index = 0
        # max distance of neibs vector from x.
        max_distance = 0
        for neib_i in range(k):
            # find max distance in this loop.
            xn = neibs[neib_i]
            
            # ||x||
            x_mag = np.dot(x, np.matrix(x).T)
            # ||xn||
            xn_mag = np.dot(xn, np.matrix(xn).T)
            
            # cosine distance between x and xn
            #d_x_xn = np.dot(x, np.matrix(xn).T) / math.sqrt(x_mag)*math.sqrt(xn_mag)
            
            # euclidean distance between x and xn
            diffxxn = x - xn
            d_x_xn = math.sqrt(np.dot(diffxxn, np.matrix(diffxxn).T))
            
            # update max value and index.
            if(d_x_xn > max_distance):
                max_distance = d_x_xn
                max_index = neib_i
        
        xd = data[data_i]
        # ||xd||
        xd_mag = np.dot(xd, np.matrix(xd).T)
        # cosine distance between x and xd
        #d_x_xd = np.dot(x, np.matrix(xd).T) / math.sqrt(x_mag)*math.sqrt(xd_mag)
        
        # euclidean distance between x and xd
        diffxxd = x - xd
        d_x_xd = math.sqrt(np.dot(diffxxd, np.matrix(diffxxd).T))
        
        # replace the most distant neighbour by the nearer data to x.
        if(max_distance > d_x_xd):
            neibs[max_index] = data[data_i]
            y_neibs[max_index] = y_data[data_i]
        
    return neibs, y_neibs
    

this gives the output of the sample requested from the dataset.

In [22]:
def supervise(n):
#     print(type(n[0]-1))
#     print('sur=',survived_vec[int(n[0]-1)])
#     print('n=',n)
    return survived_vec[int(n[0]-1)] #the first element of n is the passenger ID

because this is a classification problem we should use majority function or the mod of the K nearest data to determine what is the predicted class of the input.

In [23]:
def majority(y_neibs):
    count = [0,0]
    for n in y_neibs:
        pass
        count[int(n)] +=1
    if(count[0] >= count[1]):
        return 0;
    else:
        return 1;

In [24]:
# parameters semantics:
# data is the sample which is considered as the train data, the predictions are based on this data.
# x is the sample given to the model for new prediction.
# k is the hyper parameter of the model. which decides how many neighbours to consider for prediction.
def KNN(data, x, k, y_data):
    neibs, y_neibs = neighbours_of(data, x, k, y_data)
    prediction = majority(y_neibs)
    return prediction

# Cross-Validation 
this is for deciding the hyper parameter K in KNN algorithm.
for this matter we divide the trainset into 4 folds and validate on one fold and train on the others in each step and take average on errors to announce the validation on the parameter. we take this method when we lack training data and the test set is separated before hand.

***NOTE: we used sklearn libraries for this section***

In [25]:
np_train = trainset.to_numpy()
kf = KFold(n_splits=3)
# number of Ks checked to decide which is better based on validation set precision.
iterations = 21
# the record of all precisions calculated for each k in validation set from KNN model.
v_precision_profile = np.zeros(iterations)
for k in range(1,iterations,2):
    
    # for each precision calculated from folding validation set used by KFold library is stored in this array.
    kth_precision_vec = np.zeros(kf.get_n_splits())
    split_index = 0
    for train_i, valid_i in kf.split(np_train):
        print('split_index=', split_index)
        s_train = np_train[train_i]
        s_valid = np_train[valid_i]
        
        y_train = survived_vec[train_i]
        # not needed beucase we use the hand-written supervise function which uses the passenger ID.
        y_valid = survived_vec[valid_i]
        # prediction vector for validation set.
        p_valid = np.zeros(y_valid.shape)
        
        for p_i in range(p_valid.shape[0]):
            p_valid[p_i] = KNN(s_train, s_valid[p_i], k, y_train)
        
        kth_precision_vec[split_index] = precision(y_valid, p_valid)
        split_index += 1
        
    print('kth_precision_vec=',kth_precision_vec)
    v_precision_profile[k-1] = np.mean(kth_precision_vec)
    print('k =', k)
    print('precision =', v_precision_profile[k-1])
    print('-----------------------------------')

print('best k =',np.argmax(v_precision_profile)+1)
print(v_precision_profile)
        
        

split_index= 0
split_index= 1
split_index= 2
kth_precision_vec= [0.70707071 0.75757576 0.71043771]
k = 1
precision = 0.7250280583613916
-----------------------------------
split_index= 0
split_index= 1
split_index= 2
kth_precision_vec= [0.74074074 0.82491582 0.78451178]
k = 3
precision = 0.7833894500561168
-----------------------------------
split_index= 0
split_index= 1
split_index= 2
kth_precision_vec= [0.74747475 0.78787879 0.7979798 ]
k = 5
precision = 0.7777777777777778
-----------------------------------
split_index= 0
split_index= 1
split_index= 2
kth_precision_vec= [0.75757576 0.7979798  0.8013468 ]
k = 7
precision = 0.7856341189674523
-----------------------------------
split_index= 0
split_index= 1
split_index= 2
kth_precision_vec= [0.77104377 0.81144781 0.7979798 ]
k = 9
precision = 0.7934904601571269
-----------------------------------
split_index= 0
split_index= 1
split_index= 2
kth_precision_vec= [0.81481481 0.8047138  0.7979798 ]
k = 11
precision = 0.8058361391694725
---

so by observing the results from above we conclude that k = 13 is the best hyper parameter value for KNN model.

## Predicting test set

In [44]:
k= np.argmax(v_precision_profile)+1
np_test = testset.to_numpy()
p_test = np.zeros(np_test.shape[0])
for i in range(p_test.shape[0]):
    p_test[i] = KNN(np_train, np_test[i], k, survived_vec)
pd.DataFrame(p_test).to_csv("../testResult/knn_test.csv")

# SVM Model implemeneting using library

In [32]:
from sklearn.svm import SVC

using the same KFolding technique to announce the precision

## Using linear kernel

In [35]:
SVM = SVC(kernel='linear')
precision_vec = np.zeros(kf.get_n_splits())
split_index = 0
for train_i, valid_i in kf.split(np_train):
    print('split_index=', split_index)
    s_train = np_train[train_i]
    s_valid = np_train[valid_i]

    y_train = survived_vec[train_i]
    y_valid = survived_vec[valid_i]

    SVM.fit(s_train, y_train)
    # prediction vector for validation set.
    p_valid = SVM.predict(s_valid)



    precision_vec[split_index] = precision(y_valid, p_valid)
    split_index += 1
print('precision :', np.mean(precision_vec))

split_index= 0
split_index= 1
split_index= 2
precision : 0.7822671156004489


## Using polynomial kernel

In [37]:
SVM = SVC(kernel='poly')
precision_vec = np.zeros(kf.get_n_splits())
split_index = 0
for train_i, valid_i in kf.split(np_train):
    print('split_index=', split_index)
    s_train = np_train[train_i]
    s_valid = np_train[valid_i]

    y_train = survived_vec[train_i]
    y_valid = survived_vec[valid_i]

    SVM.fit(s_train, y_train)
    # prediction vector for validation set.
    p_valid = SVM.predict(s_valid)



    precision_vec[split_index] = precision(y_valid, p_valid)
    split_index += 1
print('precision :', np.mean(precision_vec))

split_index= 0
split_index= 1
split_index= 2
precision : 0.8047138047138048


## Using radial basis function

In [36]:
SVM = SVC(kernel='rbf')
precision_vec = np.zeros(kf.get_n_splits())
split_index = 0
for train_i, valid_i in kf.split(np_train):
    print('split_index=', split_index)
    s_train = np_train[train_i]
    s_valid = np_train[valid_i]

    y_train = survived_vec[train_i]
    y_valid = survived_vec[valid_i]

    SVM.fit(s_train, y_train)
    # prediction vector for validation set.
    p_valid = SVM.predict(s_valid)



    precision_vec[split_index] = precision(y_valid, p_valid)
    split_index += 1
print('precision :', np.mean(precision_vec))

split_index= 0
split_index= 1
split_index= 2
precision : 0.8035914702581368


## Using sigmoid kernel

In [39]:
SVM = SVC(kernel='sigmoid')
precision_vec = np.zeros(kf.get_n_splits())
split_index = 0
for train_i, valid_i in kf.split(np_train):
    print('split_index=', split_index)
    s_train = np_train[train_i]
    s_valid = np_train[valid_i]

    y_train = survived_vec[train_i]
    y_valid = survived_vec[valid_i]

    SVM.fit(s_train, y_train)
    # prediction vector for validation set.
    p_valid = SVM.predict(s_valid)



    precision_vec[split_index] = precision(y_valid, p_valid)
    split_index += 1
print('precision :', np.mean(precision_vec))

split_index= 0
split_index= 1
split_index= 2
precision : 0.6778900112233446


## SVM model conclusion

based on the above results the best model is the polynomial kernel with the default degree of 3. so we use this model to predict the testset survival.

In [43]:
SVM = SVC(kernel='poly')
SVM.fit(np_train, survived_vec)
np_test = testset.to_numpy()
svm_p_test = SVM.predict(np_test)
pd.DataFrame(svm_p_test).to_csv("../testResult/svm_test.csv")